src/matvis/gpu/matprod.py

```diff
@@ -0,0 +1,84 @@
+"""GPU-accelerated source-summing operation."""
+import cupy as cp
+import numpy as np
+
+from ..core.matprod import MatProd
+from ._cublas import dotc, zdotz
+
+
+class GPUMatMul(MatProd):
+    """Use cupy.gemm to perform the source-summing operation."""
+
+    def allocate_vis(self):
+        """Allocate memory for the visibilities.
+
+        The shape here is (nchunks, nant, nfeed, nant, nfeed), which is backwards
+        from what you'd expect (nfeed,nant, nfeed, nant). This is because the
+        fortran ordering is used in CUBLAS, which is the same as the transpose of the
+        expected shape.
+        """
+        # The shape is required to be like this to use the fortran ordering
+        self.vis = cp.full(
+            (self.nchunks, self.nant, self.nfeed, self.nant, self.nfeed),
+            0.0,
+            dtype=self.ctype,
+            order="F",
+        )
+
+    def compute(self, z: cp.ndarray, out: cp.ndarray) -> cp.ndarray:
+        """Perform the source-summing operation for a single time and chunk."""
+        zdotz(z, out=out)
+        return out
+
+    def sum_chunks(self, out: np.ndarray):
+        """Sum the chunks into the output array.
+
+        Here we need to also re-shape the visibility array into the output array.
+
+        Parameters
+        ----------
+        out
+            The output visibilities, with shape (Nfeed, Nfeed, Npairs).
+        """
+        if self.nchunks == 1:
+            cpu = self.vis[0].get()
+        else:
+            cpu = self.vis.sum(axis=0).get()
+
+        # cpu = cpu.transpose((0, 2, 1, 3))
+        cpu = cpu.transpose((1, 3, 0, 2))
+
+        if self.all_pairs:
+            cpu = cpu.reshape((self.nfeed, self.nfeed, self.nant * self.nant))
+            out[:] = cpu
+        else:
+            out[:] = cpu[:, :, self.ant1_idx, self.ant2_idx]
+
+
+class GPUVectorDot(MatProd):
+    """Use a loop over specific pairs, performing a vdot over the source axis."""
+
+    def allocate_vis(self):
+        """Allocate memory for the visibilities."""
+        self.vis = cp.full(
+            (self.nchunks, self.nfeed, self.nfeed, self.npairs), 0.0, dtype=self.ctype
+        )
+
+    def compute(self, z: cp.ndarray, out: cp.ndarray) -> cp.ndarray:
+        """Perform the source-summing operation for a single time and chunk."""
+        z = z.reshape((self.nfeed, self.nant, -1))
+
+        for j in range(self.nfeed):
+            for k in range(self.nfeed):
+                for i, (ai, aj) in enumerate(self.antpairs):
```

Comment:
per my earlier comment - with cupy I don't think the loop over nfeed is necessary